STYLE-GUIDE.md

```diff
@@ -0,0 +1,285 @@
+# Spectrometer style guide
+
+Through a number of pull requests, reviews, and discussions, a codebase style has emerged.  This document is an attempt to capture that, and prevent further discussions that slow down work unnecessarily.
+
+This style guide is forked from Github's [semantic](https://github.com/github/semantic).
+
+## General guidelines
+
+Make your code look like the code around it. Consistency is the name of the game.
+
+Use `fourmolu` for formatting.  Our CI setup enforces that all changes pass a `fourmolu` run with no differences.
+
+Our CI systems ensure that all patches pass `hlint`'s muster. We have our own set of rules in `.hlint.yaml`.
+
+We strongly recommend adding Haddock documentation to any function/data type, unless its purpose is immediately apparent from its name.
+Comments should describe the "why", type signatures should describe the "what", and the code should describe the "how".
+
+## Formatting
+
+2 spaces everywhere. Tabs are forbidden. Haskell indentation can be unpredictable, so generally stick with what your editor suggests.
+There is no hard line-length limit, though if you go beyond 110 or 120 you should generally split it up, especially for type signatures.
```

Comment:
I propose that comments should wrap at 80 (except obvious exceptions like long URLs) and lines should try to wrap at 80, but not at the expense of readability. This makes split-view code review easier on normal laptop screen sizes.

Reply:
I don't love marking off at 80, specifically.  I prefer 100/120, since 80 actually wastes a decent amount of space on my monitors.

In reality, not making a line limit is really **my** preferred option.  The formatter will do a lot of work for us here, and beyond that we can adjust it in code review.  As long as we tell people to write vertically instead of horizontally, we are probably fine.  I certainly don't want to set a 80-char limit without significant agreement from the team (i.e. majority believe that 80 is the *best* option, rather than just *an* option).

Reply:
I don't personally care what the line length is because I use wordwrap since a lot of our codebases are inconsistent but I do think we should set a hard limit. Whether it's 80, 100, 120, or nothing I think we should enforce it. I hate asking myself "is this ok?" if I have a long line and then having it be based on judgment.

Reply:
I think if you're debating whether it's okay, you should split it up.  The nice thing is that you can split it up yourself, and the formatter will handle the alignment and exact separation.  What the formatter seem to NOT do is decide when to split things up.  This is where judgement comes in.  Our tool isn't going to do this for us, so we just need to do it ourselves, and in that case, there's no benefit to setting a hard limit.

That said, I'm OK with a set of soft limits, for example:
* Write code vertically, rather than horizontally
* If it's trivial, stay under 80
* If it's reasonable, stay under 100
* If it's possible, stay under 120
* Don't go past 120 for anything except long literal strings without spaces, like URL's.
  * Prefer the cost of joining in memory over the readability problems of lines longer than 120.

Reply:
I think we should have just 1 rule, otherwise it just consumes way too much brainpower. `go fmt` has the right idea here.

The reason I think we should set the ruler at 80 is that 80 characters is the size of a half-page in the split-view code review panel on a 1920x1680 monitor (the standard size for most monitors and laptops). Otherwise, code lines will wrap for 1920x1680s (e.g. Macbooks).